File: 0948-bag-of-tokens/0948-bag-of-tokens.cpp
class Solution {
public:
    int bagOfTokensScore(vector<int>& tokens, int power) {
        sort(tokens.begin(), tokens.end()); 
        int n = tokens.size()-1; 
        int l = 0; 
        int r = n; 
        int score = 0; 
        int maxScore = 0; 


        while(l <= r){
            if(power >= tokens[l]){
                power = power - tokens[l]; 
                score++; 
                l++; 
                maxScore = max(maxScore, score); 
            }else if (score >= 1){
                power = power + tokens[r];
                score--; 
                r--; 
            }else{
                return maxScore; 
            }
        }

        return maxScore; 
    }
};

File: 0948-bag-of-tokens/0948-bag-of-tokens.py
class Solution:
    def bagOfTokensScore(self, tokens: List[int], power: int) -> int:
        tokens.sort(); 
        l = score = maxScore = 0
        r = len(tokens) -1

        while l <= r:
            if power >= tokens[l]:
                power = power - tokens[l]
                score = score + 1
                l += 1
                maxScore = max(maxScore, score); 
            elif score >= 1:
                power += tokens[r]
                score -= 1; 
                r -= 1
            else :
                return maxScore 
        return maxScore 
        